### PR TITLE
Add Uniform bounding box sampling option for CutOut and CutMix

### PR DESCRIPTION
**Pull request by coryMosaicML**
The implementation of CutOut/CutMix used in the original papers samples bounding boxes such that the probability of a given pixel being masked/mixed is not uniform across the image, with pixels near the center more likely to be masked/mixed than pixels near the edge. Our current implementations mirror this behavior.

This PR adds an additional option of sampling bounding boxes such that the probability of a given pixel being masked/mixed is uniform over the image. Default behavior remains the same as in the original papers. 

The effect of setting `uniform_sampling=True` without changing the other hyperparameters will be to increase the probability of the pixels on the edges/corners being masked/mixed, while decreasing the probability of pixels in the center being masked/mixed. This results in a smaller area of the image being masked/mixed on average, and likely implies a weaker regularizing effect, so `cutout.length` and `cutmix.alpha` may need to increase to compensate.

Test Runs with default params:
[ResNet50-i1k baseline Accuracy ~76.42%
](https://wandb.ai/corymosaicml/cory-uniform-sampling/runs/s4xlxxqt)[ResNet50-i1k CutOut default sampling Accuracy 76.67% 
](https://wandb.ai/corymosaicml/cory-uniform-sampling/runs/3grnj2ei)[ResNet50-i1k CutOut uniform sampling Accuracy 76.82% ](https://wandb.ai/corymosaicml/cory-uniform-sampling/runs/l62u6q6b)
[ResNet50-i1k CutMix default sampling Accuracy 77.63%
](https://wandb.ai/corymosaicml/cory-uniform-sampling/runs/m1aydsyt)[ResNet50-i1k CutMix uniform sampling Accuracy 77.63%](https://wandb.ai/corymosaicml/cory-uniform-sampling/runs/2ujqe24g)

### CODE REVIEW

File: composer/algorithms/cutout/cutout.py [hunk 1]
```diff
@@ -58,8 +61,12 @@ def cutout_batch(X: ImgT, n_holes: int = 1, length: Union[int, float] = 0.5) ->
 
     mask = torch.ones_like(X_tensor)
     for _ in range(n_holes):
-        y = np.random.randint(h)
-        x = np.random.randint(w)
+        if uniform_sampling is True:
```
Comment (Landanjs) [inline composer/algorithms/cutout/cutout.py:64]: Nit: don't need to check for true

**Review by Landanjs — APPROVED** [review]
Nice! LGTM, but maybe someone on research engineering could skim through too @hanlint?

I didn't trust my math checking so I made plots instead: 
image
image

Future note: it would be nice to make the code for cutmix/cutout more similar if possible for readability and any other changes

File: composer/algorithms/cutmix/cutmix.py [hunk 2]
```diff
@@ -61,6 +62,9 @@ def cutmix_batch(X: Tensor,
         bbox: optional, predetermined (rx1, ry1, rx2, ry2) coords of the bounding box.
         indices: Permutation of the batch indices `1..B`. Used
             for permuting without randomness.
+        uniform_sampling: If true, sample the bounding box such that each pixel
+            has an equal probability of being mixed. If false, defaults to the
+            sampling used in the original paper implementation.
```
Comment (hanlint) [inline composer/algorithms/cutmix/cutmix.py:67]:
```suggestion
            sampling used in the original paper implementation. Default: False
```

File: composer/algorithms/cutmix/cutmix.py [hunk 3]
```diff
@@ -61,6 +62,9 @@ def cutmix_batch(X: Tensor,
         bbox: optional, predetermined (rx1, ry1, rx2, ry2) coords of the bounding box.
         indices: Permutation of the batch indices `1..B`. Used
             for permuting without randomness.
+        uniform_sampling: If true, sample the bounding box such that each pixel
```
Comment (hanlint) [inline composer/algorithms/cutmix/cutmix.py:65]:
```suggestion
        uniform_sampling (bool, optional): If true, sample the bounding box such that each pixel
```

want to clean up these docstrings while you are at it? 😄

File: tests/algorithms/test_cutmix.py [hunk 4]
```diff
@@ -61,7 +62,8 @@ def test_cutmix(self, fake_data, alpha):
                                             H=x_fake.shape[3],
                                             cutmix_lambda=cutmix_lambda,
                                             cx=cx,
-                                            cy=cy)
+                                            cy=cy,
+                                            uniform_sampling=uniform_sampling)
         bbox = (bbx1, bby1, bbx2, bby2)
         # Adjust lambda
         cutmix_lambda = 1 - ((bbx2 - bbx1) * (bby2 - bby1) / (x_fake.size()[-1] * x_fake.size()[-2]))
```
Comment (hanlint) [inline tests/algorithms/test_cutmix.py:69]: does uniform sampling need to be passed to cutmix_batch below too?

**Review by hanlint — COMMENTED** [review]
LGTM, a few questions.

File: tests/algorithms/test_cutmix.py (hunk 4 above)
Comment (coryMosaicML) [reply tests/algorithms/test_cutmix.py:69]: Technically no, because if the explicit bounding box is given then the sampling step is skipped. I agree it looks odd though. Will fix.

**Review by A-Jacobson — APPROVED** [review]
LGTM, I don't have much to add that landen and hanlin already haven't mentioned.
